File: 10986_Sending_email/10986.java
import java.io.*;
import java.util.*;

class Main{
	public static void main(String args[]) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int Cases = Integer.parseInt(br.readLine());
		StringBuilder sb = new StringBuilder();
		int worst [] = new int[20000];
		for(int C = 1; C <= Cases; C++){
			StringTokenizer stok = new StringTokenizer(br.readLine());
			int N = Integer.parseInt(stok.nextToken()); // # of nodes
			int M = Integer.parseInt(stok.nextToken());	// # of cables
			int S = Integer.parseInt(stok.nextToken()); // start
			int T = Integer.parseInt(stok.nextToken()); // end
			HashMap<Integer, HashSet<Node> > map = new HashMap<Integer, HashSet<Node> >();			
			for(int i = 0; i < N; i++){
				worst [i] = Integer.MAX_VALUE;
				map.put(i, new HashSet<Node>());
			} // not very efficient but :/
			for(int i = 0; i < M; i++){
				stok = new StringTokenizer(br.readLine());
				int to = Integer.parseInt(stok.nextToken()); // # to
				int from = Integer.parseInt(stok.nextToken());	// # from
				int cost = Integer.parseInt(stok.nextToken()); // # cost
				map.get(to).add(new Node(cost, from));
				map.get(from).add(new Node(cost, to));
			}			
			PriorityQueue<Node> pq = new PriorityQueue<Node>(); // this requires dijkstra
			pq.add(new Node(0, S));
			worst[S] = 0;
			while(!pq.isEmpty()){
				Node node = pq.poll();
				int cur = node.current;
				int cost = node.cost;
				if(worst[cur] < cost) continue; // if you don't do this it will be really slow
				if(cur == T) break; // also needed for speed up
				for(Node other: map.get(cur)){
					int newcost = other.cost + cost;
					int temp = other.current;
					if(worst[temp] > newcost){
						worst[temp] = newcost;
						pq.add(new Node(newcost, temp));
					}
				}
			}// while more left in priority queue
			sb.append("Case #").append(C).append(": ");
			if(worst[T] == Integer.MAX_VALUE) sb.append("unreachable");
			else sb.append(worst[T]);
			sb.append('\n');
		}// while cases left
		System.out.print(sb);
	}// main
}

class Node implements Comparable<Node>{
	int cost;
	int current;
	public Node(int cost, int current){
		this.cost = cost;
		this.current = current;
	}
	public int compareTo(Node other){
		return cost - other.cost;
	}
	public int hashCode(){
		return (current << 46) | cost;
	}
	public boolean equals(Object o){
		Node other = (Node)o;
		return other.current == current && other.cost == cost;
	}
}

File: 10986_Sending_email/10986.cpp
#include<cstdio>
#include<vector>
#include<queue>
#include<map>
#include<set>
#include<utility>

using namespace std;

typedef pair<int, int> P;
typedef set<pair<int, int> >::iterator SI;
#define MP(x, y) make_pair(x,y)
#define max 2000000000

class comparator{
    public:
        bool operator() (const P &lhs, const P &rhs){
			return lhs.second > rhs.second;
		}
};

int main(){
	int worst[20000];
	int Cases;
	scanf("%d", &Cases);
	for(int C = 1; C <= Cases; C++){
		int N, M, S, T;
		scanf("%d %d %d %d", &N, &M, &S, &T);
		map<int, set<P> > m;
		for(int i = 0; i < N; i++){
			worst [i] = max;
		}
		for(int i = 0; i < M; i++){
			int to, from, cost;
			scanf("%d %d %d", &to, &from, &cost);
			m[to].insert(MP(from, cost));
			m[from].insert(MP(to, cost));
		}
		priority_queue<P, vector<P>, comparator> pq;
		worst[S] = 0;
		pq.push(MP(S, 0));
		while(!pq.empty()){
			P node = pq.top();
			pq.pop();
			int cur = node.first;
			int cost = node.second;
			if(worst[cur] < cost) continue;
			if(cur == T) break;		
			SI it = m[cur].begin();
			while(it != m[cur].end()){
				P other = *it;
				int newcost = other.second + cost;
				int temp = other.first;
				if(worst[temp] > newcost){
					worst[temp] = newcost;
					pq.push(MP(temp, newcost));
				}
				it++;
			}
		}
		printf("Case #%d: ", C);
		if(worst[T] == max) printf("unreachable");
		else printf("%d", worst[T]);
		putchar('\n');
	}// cases

}